sympy/geometry/ellipse.py

```diff
@@ -695,25 +693,41 @@ def is_tangent(self, o):
         True
 
         """
-        inter = None
-        if isinstance(o, Ellipse):
-            inter = self.intersection(o)
-            if isinstance(inter, Ellipse):
+        if isinstance(o, Point2D):
+            return False
+        elif isinstance(o, Ellipse):
+            intersect = self.intersection(o)
+            if isinstance(intersect, Ellipse):
+                return True
+            elif intersect:
+                return all((self.tangent_lines(i)[0]).equals((o.tangent_lines(i)[0])) for i in intersect)
+            else:
                 return False
-            return (inter is not None and len(inter) == 1
-                    and isinstance(inter[0], Point))
-        elif isinstance(o, LinearEntity):
-            inter = self.intersection(o)
-            if inter is not None and len(inter) == 1:
-                return inter[0] in o
+        elif isinstance(o, Line2D):
+            return len(self.intersection(o)) == 1
+        elif isinstance(o, Ray2D):
+            intersect = self.intersection(o)
+            if len(intersect) == 1:
+                return intersect[0] != o.source and not self.encloses_point(o.source)
             else:
                 return False
-        elif isinstance(o, Polygon):
-            c = 0
-            for seg in o.sides:
-                inter = self.intersection(seg)
-                c += len([True for point in inter if point in seg])
-            return c == 1
+        elif isinstance(o, (Segment2D, Polygon)):
+            any_intersect = False
```

Comment:
I recommend changing this flag to `all_tangents`. As I would understand `any_intersect`, if there any intersections, then the result should be False for `is_tangent`, but you are using the flag to keep track of whether or not there are *no* intersections.